docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,281 @@
+5. Create `LocalityLbEndpoints` with selected endpoints. 
+6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceStore`. 
+
+Dataplanes in the specific location are going to receive only endpoints in the specific locality. For cross zone traffic without egress the control-plane delivers only ingresses of specific zones.
+
+##### Egress
+Egress is not as simple as ingress. Currently, we support Locality Aware when atleast one client requires it, so we cannot distinguish between clients. Control-plane needs to send all dataplanes to the egress sidecar because there might be services sending requests to all zones. There are 2 options:
+- LoadBalacing based on metadata
+- Separate clusters
+
+Load balancing based on the metadata seems to be able to solve an issue but it has limitations. Because, egress knows all endpoints and does routing based on matching metadata the request might not obey the priority of zones. In this case we cannot configure different priorities for different clients.
+
+Separate cluster seems to be the only option that could solve this problem. When there is a different configuration of Load Balancing for the one of dataplane, control-plane sends to egress new a cluster `mesh-name:service-name:hash(tag + value)` with the routing based on the SNI: `service-name{mesh=mesh-name,hash=hash}`
+This cluster has a different configuration of priorities and only subset of zones. Dataplanes interested in this Load Balancing are going to have the same configuration of the cluster which sets SNI: `service-name{mesh=mesh-name,hash=hash}` and based in this SNI egress can make a routing decision to route to the specific cluster.
+
+##### Ingress
+Ingress should receive all endpoints, because there is no rules for in zone traffic when the request comes from the outside.
+
+##### Serviceless
+On kubernetes you can create a deployment without the service, in this case control-planes creates default inbound so tags should be available.
+On universal you cannot create a dataplane without inbound or gateway definition.
+
+#### ClusterLoadBalancing cache
+Currently, while retrieving endpoints in the plugin, we are getting a pointer to the object in the cache that is shared between dataplanes. This won't work with LocalityAware load balancing because each dataplane might have a different subset of endpoints with different priorities. In the first iteration, we would like to try using it without CLA cache and measure the impact of the policy in the bigger cluster with many nodes. If the impact of not using the cache is significant, we might have to keep this configuration in the cache if there are more services on the same node. The cache might not improve the performance if there are instances of the same service in different localities (nodes, availability zones).
```

Comment:
@lobkovilya and @jakubdyszkiewicz can probably already tell you that we can't live without the CLA cache and that's why it was introduced.

However the CLA cache key is:

`key := sha256.Hash(fmt.Sprintf("%s:%s:%s:%s", apiVersion, meshName, cluster.Hash(), meshHash))`

You could quite easily keep a CLA per `rule` and still have the cache.

I think picking the right rule is probably pretty quick (especially as in most cases there will only be a top level targetRef of `kind:Mesh`)

Reply:
It's hard to say if we really can't live without CLA cache. That was the first optimization for CP perf even before MeshCache and way way before reachable services. If you use reachable services, I'm not sure if the CLA cache is that critical. We need to measure it.

Reply:
Maybe I'm reading it wrong, but I think your cache key wouldn't work, because you need to also add "input for your load balancing policy" to the key. For example. You may have same policy to prefer `k8s.io/node`, but DPP on different nodes would have CLA with different priorities.

Reply:
we will deal with CLA cache later on as a separate issue